web/webapi.py

```diff
@@ -423,47 +406,60 @@ def rawinput(method=None):
     method = method or "both"
 
     def dictify(fs):
-        # hack to make web.input work with enctype='text/plain.
-        if fs.list is None:
-            fs.list = []
-
         return {k: fs[k] for k in fs}
 
-    e = ctx.env.copy()
+    env = ctx.env.copy()
     a = b = {}
 
+    # This is required by the multipart module and not set up by the tests, so add the default value if missing
+    # srb: not sure if thtis is needed, as OpenLibrary appears to work without it. It also changes the content
+    #      type for many requests. Change tests, or add this? Not sure.
+    # if "CONTENT_TYPE" not in env:
+    #     env["CONTENT_TYPE"] = "application/x-www-form-urlencoded"
+
     if method.lower() in ["both", "post", "put", "patch"]:
-        if e["REQUEST_METHOD"] in ["POST", "PUT", "PATCH"]:
-            if e.get("CONTENT_TYPE", "").lower().startswith("multipart/"):
-                # since wsgi.input is directly passed to cgi.FieldStorage,
-                # it can not be called multiple times. Saving the FieldStorage
+        if env["REQUEST_METHOD"] in ["POST", "PUT", "PATCH"]:
+            if env.get("CONTENT_TYPE", "").lower().startswith("multipart/"):
+                # since wsgi.input is directly passed to multipart,
+                # it can not be called multiple times. Saving the result
                 # object in ctx to allow calling web.input multiple times.
-                a = ctx.get("_fieldstorage")
+                a = ctx.get(
+                    "_fieldstorage"
+                )  # TODO: Rename? is this visible anywhere else?
                 if not a:
-                    fp = e["wsgi.input"]
-                    a = cgiFieldStorage(fp=fp, environ=e, keep_blank_values=1)
-                    ctx._fieldstorage = a
+                    try:
+                        # This returns two dicts, forms & files.
+                        forms, a = multipart.parse_form_data(environ=env)
+                        a = dictadd(forms, a)
+                        ctx._fieldstorage = a
+                    except IndexError:
+                        a = {}
+
             else:
-                d = data()
-                if isinstance(d, str):
-                    d = d.encode("utf-8")
-                fp = BytesIO(d)
-                a = cgiFieldStorage(fp=fp, environ=e, keep_blank_values=1)
+                post_data = data().decode("utf-8")
```

Comment:
Just making sure we're not accidentally losing anything in this `decode` (or hitting any other unexpected results)